contracts/upgradeable_contracts/erc20_to_native/BridgeRouter.sol

```diff
@@ -0,0 +1,115 @@
+    address public immutable FOREIGN_XDAIBRIDGE = 0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016;
+    address public immutable DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+    address public immutable USDS = 0xdC035D45d973E3EC169d2276DDab16f1e407384F;
+    address public immutable WETH_OMNIBRIDGE_ROUTER = 0xa6439Ca0FCbA1d0F80df0bE6A17220feD9c9038a;
+    uint256[49] __gap;
+
+    mapping(address => address) public tokenRoutes;
+
+    function initialize() public initializer {
+        __Ownable_init(msg.sender);
+    }
+
+    /// @notice An entry point contract for user to bridge any token from source chain
+    /// @dev Directs route to relevant contract to perform token relaying
+    /// @param _token token to bridge
+    /// @param _receiver receiver of token on Gnosis Chain
+    /// @param _amount amount to receive on Gnosis Chain
+    function relayTokens(address _token, address _receiver, uint256 _amount) external payable {
+        address route = tokenRoutes[_token];
+
+        if (_token == DAI) {
+            IERC20(_token).transferFrom(msg.sender, route, _amount);
+            IXDaiBridgePeripheral(route).relayTokens(_receiver, _amount);
+        } else if (_token == USDS) {
+            // token need to be transferred to router contract first, because the bridge will call transferFrom(msg.sender, bridge, amount);
+            IERC20(_token).transferFrom(msg.sender, address(this), _amount);
+            IERC20(_token).approve(route, _amount);
+            IXDaiBridgePeripheral(route).relayTokens(_receiver, _amount);
+        } else if(_token == address(0)){
+            // call wrapAndRelayTokens
+            require(msg.value == _amount, "msg.value mismatch");
+            IWETHOmnibridgeRouter(WETH_OMNIBRIDGE_ROUTER).wrapAndRelayTokens{value: msg.value}(_receiver);
+        }   
+        else {
+            IERC20(_token).transferFrom(msg.sender, address(this), _amount);
+            IERC20(_token).approve(FOREIGN_OMNIBRIDGE, _amount);
+            IForeignBridge(FOREIGN_OMNIBRIDGE).relayTokens(_token, _receiver, _amount);
+        }
+    }
+
+    /// @notice Set route for specific token
+    /// @param _token token address
+    /// @param _route router contract address
+    function setRoute(address _token, address _route) public onlyOwner {
+        require(_route != address(0) && _route != address(this), "invalid route address");
+        uint256 size;
+        assembly {
+            size := extcodesize(_route)
+        }
+        require(size > 0, "route should be a contract");
+        tokenRoutes[_token] = _route;
+    }
+
+    /// @notice Claim token function
+    /// @dev This function check if the data belongs of xDAI bridge or AMB/Omnibridge
+    /// @param message for claiming tx
+    /// @param signatures signatures from bridge validators
+    function executeSignatures(bytes memory message, bytes memory signatures) external {
+        if (message.length == 104) {
```

Comment:
Are we sure that an AMB message cannot have a length of 104?

Reply:
AMB should be at least 111 bytes I believe: https://github.com/gnosischain/tokenbridge-contracts/blob/812356dd322ba0bd4e0d0b7f007737e2865f9320/contracts/libraries/ArbitraryMessage.sol#L7